File: hackerrank/problem solving/algorithms/2 - implementation/36 - repeated string/oo/repeated_string_oo.rb
# Source: https://www.hackerrank.com/challenges/repeated-string/problem?isFullScreen=true

def main
    inputString = gets.rstrip
    nCharacters = gets.to_i

    obj = RepeatedString.new inputString, nCharacters
    quantityOfA = obj.getQuantityOfA
    puts quantityOfA
end

    class RepeatedString
        @inputString
        @nCharacters
        @quantityOfA

        def initialize inputString, nCharacters
            @inputString = inputString
            @nCharacters = nCharacters
            @quantityOfA = 0

            _findQuantityOfAInRepeatedString
        end

            def _findQuantityOfAInRepeatedString
                @quantityOfA = _findQuantityOfAInString @inputString
                @quantityOfA *= _findQuantityOfRepeatedEntireString
                @quantityOfA += _findQuantityOfRemainingA
            end

                def _findQuantityOfAInString inputString
                    quantityOfA = 0

                    for character in inputString.chars
                        if character == 'a'
                            quantityOfA += 1
                        end
                    end

                    quantityOfA
                end

                def _findQuantityOfRepeatedEntireString
                    nStrings = (@nCharacters / @inputString.size).truncate
                end

                def _findQuantityOfRemainingA
                    nRemainingCharactersOfString = @nCharacters % @inputString.size

                    if nRemainingCharactersOfString == 0
                        return 0
                    end

                    subString = @inputString[0..(nRemainingCharactersOfString-1)]
                    return _findQuantityOfAInString subString
                end

        def getQuantityOfA
            @quantityOfA
        end
    end

main

File: hackerrank/problem solving/algorithms/2 - implementation/36 - repeated string/oo/repeated_string_oo.cpp
// Source: https://www.hackerrank.com/challenges/repeated-string/problem?isFullScreen=true

#include <iostream>
#include <string>

using namespace std;

string read_a_string();
unsigned long read_a_number();


class Repeated_String {

private:
    string _input_string;
    unsigned long _n_characters;
    unsigned long _quantity_of_a;

    void _find_quantity_of_a_in_repeated_string() {
        _quantity_of_a = {_find_quantity_of_a_in_string(_input_string)};
        _quantity_of_a *= _find_quantity_of_repeated_entire_string();
        _quantity_of_a += _find_quantity_of_remaining_a();
    }

        unsigned long _find_quantity_of_a_in_string(const string INPUT_STRING) const {
            unsigned long quantity_of_a {0};

            for (auto character : INPUT_STRING)
                if (character == 'a')
                    quantity_of_a++;

            return quantity_of_a;
        }

        unsigned long _find_quantity_of_repeated_entire_string() const {
            const unsigned long N_STRINGS {_n_characters / _input_string.length()};
            return N_STRINGS;
        }

        unsigned long _find_quantity_of_remaining_a() const {
            const unsigned long N_REMAINING_CHARACTERS_OF_STRING {_n_characters % _input_string.length()};

            if (N_REMAINING_CHARACTERS_OF_STRING == 0)
                return 0;

            const string SUB_STRING {_input_string.substr(0, N_REMAINING_CHARACTERS_OF_STRING)};
            return _find_quantity_of_a_in_string(SUB_STRING);
        }

public:
    Repeated_String(const string INPUT_STRING, const unsigned long N_CHARACTERS) {
        _input_string = {INPUT_STRING};
        _n_characters = {N_CHARACTERS};
        _quantity_of_a = {0};

        _find_quantity_of_a_in_repeated_string();
    }

    unsigned long get_quantity_of_a() const {
        return _quantity_of_a;
    }
};

int main() {
    const string INPUT_STRING {read_a_string()};
    const unsigned long N_CHARACTERS {read_a_number()};

    const Repeated_String OBJ(INPUT_STRING, N_CHARACTERS);
    const unsigned long QUANTITY_OF_A {OBJ.get_quantity_of_a()};
    cout << QUANTITY_OF_A << "\n";

    return 0;
}

    string read_a_string() {
        string input_string;
        cin >> input_string;
        return input_string;
    }

    unsigned long read_a_number() {
        unsigned long number;
        cin >> number;
        return number;
    }
